contracts/treasury/src/handle.rs

```diff
@@ -1,43 +1,119 @@
+use cosmwasm_std;
 use cosmwasm_std::{
-    debug_print,
-    to_binary,
-    Api,
-    Binary,
-    Env,
-    Extern,
-    HandleResponse,
-    HumanAddr,
-    Querier,
-    StdError,
-    StdResult,
-    Storage,
-    Uint128,
+    from_binary, to_binary, Api, Binary, Env, Extern, HandleResponse, HumanAddr, Querier, StdError,
+    StdResult, Storage, Uint128,
 };
-use secret_toolkit::snip20::{register_receive_msg, set_viewing_key_msg, token_info_query};
+use secret_toolkit::snip20::{register_receive_msg, send_msg, set_viewing_key_msg};
 
 use shade_protocol::{
     asset::Contract,
     generic_response::ResponseStatus,
-    treasury::{Allocation, Asset, HandleAnswer},
+    //math::Uint128,
+    snip20::fetch_snip20,
+    treasury::{Allocation, Config, Flag, HandleAnswer, QueryAnswer},
 };
 
-use crate::state::{assets_r, assets_w, config_r, config_w, viewing_key_r};
+use crate::{
+    query,
+    state::{allocations_w, assets_r, assets_w, config_r, config_w, reserves_w, viewing_key_r},
+};
 
 pub fn receive<S: Storage, A: Api, Q: Querier>(
     deps: &mut Extern<S, A, Q>,
     env: Env,
     _sender: HumanAddr,
     _from: HumanAddr,
     amount: Uint128,
-    _msg: Option<Binary>,
+    msg: Option<Binary>,
 ) -> StdResult<HandleResponse> {
-    let assets = assets_r(&deps.storage);
+    let asset = assets_r(&deps.storage).load(env.message.sender.to_string().as_bytes())?;
+    //debug_print!("Treasured {} u{}", amount, asset.token_info.symbol);
+    // skip the rest if the send the "unallocated" flag
+    if let Some(f) = msg {
+        let flag: Flag = from_binary(&f)?;
+        if flag.flag == "unallocated" {
+            return Ok(HandleResponse {
+                messages: vec![],
+                log: vec![],
+                data: Some(to_binary(&HandleAnswer::Receive {
+                    status: ResponseStatus::Success,
+                })?),
+            });
+        }
+    };
+
+    let mut messages = vec![];
+
+    allocations_w(&mut deps.storage).update(
+        asset.contract.address.to_string().as_bytes(),
+        |allocs| {
+            let mut alloc_list = match allocs {
+                None => {
+                    vec![]
+                }
+                Some(a) => a,
+            };
```

Comment:
You can clean up all of the matches like this

```suggestion
            let mut alloc_list = allocs.unwrap_or(vec![]);
```